File: pyspark/train.py
"""
PySpark Decision Tree Regression Example.
"""

from pyspark.sql import SparkSession
from pyspark.ml import Pipeline
from pyspark.ml.regression import DecisionTreeRegressor
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import VectorAssembler
import mlflow
import mlflow.spark
from common import *

spark = SparkSession.builder.appName("App").getOrCreate()

print("MLflow Version:", mlflow.version.VERSION)
print("Spark Version:", spark.version)
print("Tracking URI:", mlflow.tracking.get_tracking_uri())

metrics = ["rmse", "r2", "mae"]

def train(data, maxDepth, maxBins, run_id, log_as_onnx):
    (trainingData, testData) = data.randomSplit([0.7, 0.3], 2019)
    print("testData.schema:")
    testData.printSchema()

    # MLflow - log parameters
    print("Parameters:")
    print("  maxDepth:", maxDepth)
    print("  maxBins:", maxBins)
    mlflow.log_param("maxDepth", maxDepth)
    mlflow.log_param("maxBins", maxBins)

    # Create pipeline
    dt = DecisionTreeRegressor(labelCol=colLabel, featuresCol=colFeatures, maxDepth=maxDepth, maxBins=maxBins)
    assembler = VectorAssembler(inputCols=data.columns[:-1], outputCol=colFeatures)
    pipeline = Pipeline(stages=[assembler, dt])
    
    # Fit model and predict
    model = pipeline.fit(trainingData)
    predictions = model.transform(testData)

    # MLflow - log metrics
    print("Metrics:")
    predictions = model.transform(testData)
    for metric_name in metrics:
        evaluator = RegressionEvaluator(labelCol=colLabel, predictionCol=colPrediction, metricName=metric_name)
        metric_value = evaluator.evaluate(predictions)
        print(f"  {metric_name}: {metric_value}")
        mlflow.log_metric(metric_name,metric_value)

    # MLflow - log spark model
    mlflow.spark.log_model(model, "spark-model")

    # MLflow - log mleap model
    mleapData = testData.drop("quality")
    mlflow.mleap.log_model(spark_model=model, sample_input=mleapData, artifact_path="mleap-model")

    # Log mleap schema file for MLeap runtime deserialization
    schema_path = "schema.json"
    with open(schema_path, 'w') as f:
        f.write(mleapData.schema.json())
    print("schema_path:", schema_path)
    mlflow.log_artifact(schema_path, "mleap-model")

    # MLflow - log onnx model
    if log_as_onnx:
        import onnx_utils
        onnx_utils.log_model(spark, model, "onnx-model", mleapData)

if __name__ == "__main__":
    from argparse import ArgumentParser
    parser = ArgumentParser()
    parser.add_argument("--experiment_name", dest="experiment_name", help="Experiment _name", default="pyspark")
    parser.add_argument("--data_path", dest="data_path", help="Data path", default=default_data_path)
    parser.add_argument("--max_depth", dest="max_depth", help="Max depth", default=5, type=int) # per doc
    parser.add_argument("--max_bins", dest="max_bins", help="Max bins", default=32, type=int) # per doc
    parser.add_argument("--describe", dest="describe", help="Describe data", default=False, action='store_true')
    parser.add_argument("--log_as_onnx", dest="log_as_onnx", help="Log model as ONNX", default=False, action='store_true')
    args = parser.parse_args()
    print("Arguments:")
    for arg in vars(args):
        print(f"  {arg}: {getattr(args, arg)}")

    client = mlflow.tracking.MlflowClient()
    mlflow.set_experiment(args.experiment_name)

    data_path = args.data_path or default_data_path
    data = read_data(spark, data_path)
    if (args.describe):
        print("==== Data")
        data.describe().show()

    with mlflow.start_run() as run:
        print("MLflow:")
        print("  run_id:", run.info.run_id)
        print("  experiment_id:", run.info.experiment_id)
        print("  experiment_name:", client.get_experiment(run.info.experiment_id).name)
        mlflow.set_tag("mlflow_version", mlflow.version.VERSION)
        mlflow.set_tag("spark_version", spark.version)
        train(data, args.max_depth, args.max_bins, run.info.run_id, args.log_as_onnx)

File: scala/sparkml/src/main/scala/org/andre/mlflow/examples/wine/xgboost/Train.scala
package org.andre.mlflow.examples.wine.xgboost

import java.io.File
import com.beust.jcommander.{JCommander, Parameter}
import org.apache.spark.sql.SparkSession
import ml.dmlc.xgboost4j.scala.spark.XGBoostRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.mlflow.tracking.{MlflowClient,MlflowClientVersion}
import org.mlflow.api.proto.Service.RunStatus
import org.andre.mlflow.util.MLflowUtils
import org.andre.mlflow.examples.wine.WineUtils

object Train {
  val spark = SparkSession.builder.appName("Train").getOrCreate()
  MLflowUtils.showVersions(spark)

  def main(args: Array[String]) {
    new JCommander(opts, args: _*)
    println("Options:")
    println(s"  trackingUri: ${opts.trackingUri}")
    println(s"  token: ${opts.token}")
    println(s"  experimentName: ${opts.experimentName}")
    println(s"  dataPath: ${opts.dataPath}")
    println(s"  modelPath: ${opts.modelPath}")
    println(s"  maxDepth: ${opts.maxDepth}")

    // MLflow - create or get existing experiment
    val client = MLflowUtils.createMlflowClient(opts.trackingUri, opts.token)

    val experimentId = MLflowUtils.getOrCreateExperimentId(client, opts.experimentName)
    println("Experiment ID: "+experimentId)

    // Train model
    train(client, experimentId, opts.modelPath, opts.maxDepth, opts.objective, opts.dataPath)
  }

  def train(client: MlflowClient, experimentId: String, modelPath: String, maxDepth: Int, objective: String, dataPath: String) {
    // Read data
    val data = WineUtils.readData(spark, dataPath)

    // MLflow - create run
    val runInfo = client.createRun(experimentId)
    val runId = runInfo.getRunId()
    println(s"Run ID: $runId")
    println(s"sparkVersion: ${spark.version}")
    println(s"scalaVersion: ${util.Properties.versionString}")

    // MLflow - set tags
    client.setTag(runId, "dataPath",dataPath)
    client.setTag(runId, "mlflow.source.name",MLflowUtils.getSourceName(getClass()))
    client.setTag(runId, "mlflowVersion",MlflowClientVersion.getClientVersion())
    client.setTag(runId, "sparkVersion",spark.version)
    client.setTag(runId, "scalaVersion",util.Properties.versionString)

    // Process data
    val dataTransformed = Utils.transformData(data)
    println("dataTransformed.schema: "+ dataTransformed.schema.treeString)
    val Array(train, eval1, eval2, test) = dataTransformed.randomSplit(Array(0.6, 0.2, 0.1, 0.1))

    // Parameters
    val params = Map("eta" -> "1", 
      "max_depth" -> maxDepth, 
      "objective" -> "rank:pairwise",
      "num_round" -> 5, 
      "num_workers" -> 2, 
      "objective" -> objective,
      "eval_sets" -> Map("eval1" -> eval1, "eval2" -> eval2))
    println("Parameters:")
    for ((k,v) <- params) {
      println(s"  $k: $v")
      client.logParam(runId, k, v.toString)
    }

    val xgb = new XGBoostRegressor(params)
      .setFeaturesCol(WineUtils.colFeatures)
      .setLabelCol(Utils.colOutput)
    val model = xgb.fit(train)
    val predictions = model.transform(test)
    println("Predictions:")
    predictions.show(5)

    // MLflow - log metrics
    val metrics = Seq("rmse","r2", "mae")
    println("Metrics:")
    for (metric <- metrics) {
      val evaluator = new RegressionEvaluator()
        .setLabelCol(Utils.colOutput)
        .setPredictionCol(WineUtils.colPrediction)
        .setMetricName(metric)
      val v = evaluator.evaluate(predictions)
      println(s"  $metric: $v - isLargerBetter: ${evaluator.isLargerBetter}")
      client.logMetric(runId, metric, v)
    }

    // MLflow - log model
    val path = s"$modelPath/spark-model"
    model.write.overwrite().save(path)
    client.logArtifacts(runId, new File(path), "xgboost-model")

    // MLflow - close run
    client.setTerminated(runId, RunStatus.FINISHED, System.currentTimeMillis())
  }

  object opts {
    @Parameter(names = Array("--trackingUri" ), description = "Tracking Server URI", required=false)
    var trackingUri: String = null

    @Parameter(names = Array("--token" ), description = "REST API token", required=false)
    var token: String = null

    @Parameter(names = Array("--dataPath" ), description = "Data path", required=true)
    var dataPath: String = null

    @Parameter(names = Array("--modelPath" ), description = "Model path", required=true)
    var modelPath: String = null

    @Parameter(names = Array("--maxDepth" ), description = "maxDepth param", required=false)
    var maxDepth: Int = 6 // per doc

    @Parameter(names = Array("--objective" ), description = "objective param", required=false)
    var objective = "reg:squarederror" // per doc

    @Parameter(names = Array("--experimentName" ), description = "Experiment name", required=false)
    var experimentName = "scala_classic"
  }
}
